### PR TITLE
C# bindings: Decode exception messages as UTF-8 strings

### PR DESCRIPTION
**Pull request by Mbucari**
## What does this PR do?

Enable UTF-8 encoded exception messages. Replaces SWIG's built-in `SWIGExceptionHelper` with `ExceptionHelperUtf8`, which decodes exception message strings with UTF-8 instead of converting them to ANSI.

It also disables the unused SWIG "string with length helper".

This will conflict with changes in #14366, and this PR should be merged first.

@runette

## What are related issues/pull requests?

Building on the changes made in #14283.

## AI tool usage

 - [ ] AI (Y-a-t-il-un-Copilot-dans-l'avion, Chat-j'ai-pété, Jean-Claude Dusse or something similar) supported my development of this PR. See our [policy about AI tool use](https://gdal.org/community/ai_tool_policy.html). Use of AI tools *must* be indicated.

## Tasklist

 - [X] Make sure code is correctly formatted (cf [pre-commit configuration](https://gdal.org/development/dev_practices.html#commit-hooks))
 - [X] Add test case(s)
 - [x] Review
 - [x] Adjust for comments

## Environment

Provide environment details, if relevant:
Mono and dotnet.

### CODE REVIEW

**Comment by runette** [comment]
@Mbucari Can you rebase aginast the update Cmake build scripts in master and rerun the tests.

That is what I was was waiting for

**Comment by rouault** [comment]
@runette Please approve if OK for you

File: swig/include/csharp/csharp_strings.i [hunk 1]
```diff
@@ -358,3 +358,108 @@ SWIGEXPORT void SWIGSTDCALL RegisterUtf8StringCallback_$module(CSharpUtf8StringH
...
+      => SWIGPendingException.Set(new OutOfMemoryException(Utf8UnmanagedToString(pMessage), SWIGPendingException.Retrieve()));
+    static void SetPendingOverflowException(IntPtr pMessage)
+      => SWIGPendingException.Set(new OverflowException(Utf8UnmanagedToString(pMessage), SWIGPendingException.Retrieve()));
+    static void SetPendingSystemException(IntPtr pMessage)
+      => SWIGPendingException.Set(new SystemException(Utf8UnmanagedToString(pMessage), SWIGPendingException.Retrieve()));
+    static void SetPendingArgumentException(IntPtr pMessage, IntPtr pParamName)
+      => SWIGPendingException.Set(new ArgumentException(Utf8UnmanagedToString(pMessage), Utf8UnmanagedToString(pParamName), SWIGPendingException.Retrieve()));
+    static void SetPendingArgumentNullException(IntPtr pMessage, IntPtr pParamName) {
+      Exception e = SWIGPendingException.Retrieve();
+	  string message = e == null ? " Inner Exception: " + e.Message : Utf8UnmanagedToString(pMessage);
```
Comment (runette) [inline swig/include/csharp/csharp_strings.i]: I think the comparison is the wrong way around -> if e == null then e.Message is a null reference ??

Also - won't this lose the stack trace ?

File: swig/include/csharp/csharp_strings.i [hunk 2]
```diff
@@ -358,3 +358,108 @@ SWIGEXPORT void SWIGSTDCALL RegisterUtf8StringCallback_$module(CSharpUtf8StringH
...
+      SWIGPendingException.Set(new ArgumentNullException(Utf8UnmanagedToString(pParamName), message));
+    }
+    static void SetPendingArgumentOutOfRangeException(IntPtr pMessage, IntPtr pParamName) {
+      Exception e = SWIGPendingException.Retrieve();
+	  string message = e == null ? " Inner Exception: " + e.Message : Utf8UnmanagedToString(pMessage);
```
Comment (runette) [inline swig/include/csharp/csharp_strings.i]: As above

File: swig/include/csharp/csharp_strings.i [hunk 3]
```diff
@@ -358,3 +358,108 @@ SWIGEXPORT void SWIGSTDCALL RegisterUtf8StringCallback_$module(CSharpUtf8StringH
...
+        new ExceptionDelegateUtf8(SetPendingSystemException));
+		
+      RegisterExceptionCallbacksArgumentUtf8(
+        new ExceptionArgumentDelegateUtf8(SetPendingArgumentException),
+        new ExceptionArgumentDelegateUtf8(SetPendingArgumentNullException),
+        new ExceptionArgumentDelegateUtf8(SetPendingArgumentOutOfRangeException));
+    }
+
+    static void SetPendingApplicationException(IntPtr pMessage)
+      => SWIGPendingException.Set(new ApplicationException(Utf8UnmanagedToString(pMessage), SWIGPendingException.Retrieve()));
```
Comment (runette) [inline swig/include/csharp/csharp_strings.i:419]: I am not clear on what happend if UTF8UnmanagedToString() throws an exception - for instance invalid UTF8 string.

File: swig/include/csharp/csharp_strings.i [hunk 4]
```diff
@@ -358,3 +358,108 @@ SWIGEXPORT void SWIGSTDCALL RegisterUtf8StringCallback_$module(CSharpUtf8StringH
...
+    }
+    unsafe static string Utf8UnmanagedToString(IntPtr pUtf8Bts) {
+      if (pUtf8Bts == IntPtr.Zero)
+        return null;
+
+      byte* pStringUtf8 = (byte*)pUtf8Bts;
+      int len = 0;
+      while (pStringUtf8[len] != 0) len = checked(len + 1);
+
+      return System.Text.Encoding.UTF8.GetString(pStringUtf8, len);
```
Comment (runette) [inline swig/include/csharp/csharp_strings.i:460]: I am worried about what state we are in if this block throws

**Review by runette — CHANGES_REQUESTED** [review]
This is very intrusive into the current code and, whilst I understand the need, I am concerned about the number of edge and corner cases this will throw up.

I do not think that current testing has been through most of the possible error paths. At the very least - we need a test that throws each of the error types to make sure that they are all working. The whole area of "what happens if the new callbacks throw" needs investigation as well

File: swig/include/csharp/csharp_strings.i (hunk 1 above)
Comment (Mbucari) [reply swig/include/csharp/csharp_strings.i]: > I think the comparison is the wrong way around -> if e == null then e.Message is a null reference ??

You're right. I pushed a fix.

> Also - won't this lose the stack trace ?

Yes, but this is how [SWIG does it](https://github.com/swig/swig/blob/226cebfb9892aae5e27f39911a035232f2275867/Lib/csharp/csharphead.swg#L209-L218). There's no `ArgumentOutOfRangeException` that accepts both a parameter name and an inner exception.

File: swig/include/csharp/csharp_strings.i (hunk 4 above)
Comment (Mbucari) [reply swig/include/csharp/csharp_strings.i:460]: The only way this block throws is with an int32 overflow when searching for a null terminator. As I stated above, `GetString` won't throw.

EDIT* See [my below comment](https://github.com/OSGeo/gdal/pull/14474#issuecomment-4415989677) for an explanation of what happens if this method throws).

File: swig/include/csharp/csharp_strings.i (hunk 3 above)
Comment (Mbucari) [reply swig/include/csharp/csharp_strings.i:419]: `GetString` won't throw because the [UTF8Encoding](https://learn.microsoft.com/en-us/dotnet/api/system.text.utf8encoding?view=net-10.0) instance accessed through `Encoding.UTF8` was not created with the [throwOnInvalidBytes](https://learn.microsoft.com/en-us/dotnet/api/system.text.utf8encoding.-ctor?view=net-10.0#system-text-utf8encoding-ctor(system-boolean-system-boolean)) option.

**Comment by Mbucari** [comment]
> At the very least - we need a test that throws each of the error types to make sure that they are all working.

Since Gdal's C# bindings currently only use SWIG_RuntimeError, I'll need to add a new test method that calls SWIG_CSharpException for every exception type. That new function will only exist for testing purposes and shouldn't be used in production, so how should I go about writing that? Maybe write a private test method inside %imclasscode and call that test method via reflection from within the test app?

> The whole area of "what happens if the new callbacks throw" needs investigation as well

Exceptions are set from exactly one place: [the global exception handler](https://github.com/OSGeo/gdal/blob/4154030f36cbf0d24bf52443392c628d3a2f91a5/swig/include/cpl_exceptions.i#L46). That handler calls `SWIG_CSharpException` after the native call `$action` has been completed but before returning to managed code. The delegate throwing an exception while trying to set the managed exception leaves GDAL in the same state as if the exception had been successfully set and then thrown. Additionally, the stack successfully unwinds in both cases.

Here's what happens when an ApplicationException is successfully thrown (by calling `Gdal.Error(CPLErr.CE_Failure, 0, "Test Error");`)
```
System.ApplicationException: Test Error
   at OSGeo.GDAL.Gdal.Error(CPLErr msg_class, Int32 err_code, String msg) in C:\dev\gdal\build\swig\csharp\gdal\Gdal.cs:line 217
   at ConsoleApp2.Program.Main(String[] args) in C:\Users\mbuca\source\repos\ConsoleApp2\ConsoleApp2\Program.cs:line 115
```

And here's what happens when Utf8UnmanagedToString throws (by hard-coding an exception into Utf8UnmanagedToString)

```
System.Exception: Utf8UnmanagedToString threw
   at OSGeo.GDAL.GdalPINVOKE.ExceptionHelperUtf8.Utf8UnmanagedToString(IntPtr pUtf8Bts) in C:\dev\gdal\build\swig\csharp\gdal\GdalPINVOKE.cs:line 379
   at OSGeo.GDAL.GdalPINVOKE.ExceptionHelperUtf8.SetPendingApplicationException(IntPtr pMessage) in C:\dev\gdal\build\swig\csharp\gdal\GdalPINVOKE.cs:line 337
   at OSGeo.GDAL.GdalPINVOKE.Error(Int32 jarg1, Int32 jarg2, IntPtr jarg3)
   at OSGeo.GDAL.Gdal.Error(CPLErr msg_class, Int32 err_code, String msg) in C:\dev\gdal\build\swig\csharp\gdal\Gdal.cs:line 216
   at ConsoleApp2.Program.Main(String[] args) in C:\Users\mbuca\source\repos\ConsoleApp2\ConsoleApp2\Program.cs:line 115
```

**Comment by Mbucari** [comment]
> we need a test that throws each of the error types to make sure that they are all working.

Added in aab58d52925c2dd84bd8baaa68eb5652bc8e337a

I exposed 2 methods, `TestSwigSetException` and `TestSwigSetArgumentException`, which are private to the `Gdal`, and the test app uses reflection to call those methods. The tests:
- Set exceptions for all SWIG-defined exception codes
- Set exceptions for undefined exception codes (int.MinValue and int.MaxValue)
- Verify that the correct exception is thrown
- Verify that the unicode strings in exception messages (and parameter names in argument exceptions) are correctly decoded.

**Review by runette — APPROVED** [review]
